syntax = "proto3";

package atc;

service ATCService {
    rpc RequestLanding (LandingRequest) returns (ATCResponse);
    rpc RequestTakeoff (TakeoffRequest) returns (ATCResponse);
    rpc CreateAircraft (Aircraft) returns (StatusUpdate);
    rpc UpdateStatus (StatusUpdate) returns (ATCResponse);
    rpc GetSystemStatus (StatusRequest) returns (SystemStatus);
    rpc ToggleGenerator (GeneratorRequest) returns (ATCResponse);
}

message LandingRequest {
    string flight_number = 1;
    string aircraft_type = 2;
    int32 size = 3;  // 1=SMALL, 2=MEDIUM, 3=LARGE
}

message TakeoffRequest {
    string flight_number = 1;
    string aircraft_type = 2;
    int32 size = 3;
}

message StatusUpdate {
    string flight_number = 1;
    string status = 2;  //LANDED, DEPARTED
}

message ATCResponse {
    string flight_number = 1;
    string instruction = 2;  //hold, go_around, cleared_for_landing
    int32 assigned_runway = 3;
    int32 estimated_wait_time = 4;
}

message StatusRequest {}

message SystemStatus {
    repeated Aircraft aircrafts = 1;
    repeated Runway runways = 2;
    repeated string landing_queue = 3;
    repeated string takeoff_queue = 4;
    repeated string logs = 5;
}

message Aircraft {
    string flight_number = 1;
    int32 size = 2;
    string status = 3;
    string requested_operation = 4;
    int32 runway_assigned = 5;
    int32 delay = 6;
    string timestamp = 7;
}

message Runway {
    int32 id = 1;
    int32 length = 2;
    string status = 3;
    string current_operation = 4;
    string current_aircraft = 5;
}

message GeneratorRequest {
    string action = 1;  //start stop
}